[FLANG] [DOCS] Update current status details of DO CONCURRENT support (#194582)
llvm/llvm-project: flang/docs/DoConcurrentConversionToOpenMP.md

## Current status

> As of LLVM 22, Flang has better support for do concurrent parallelization on the CPU

Reviewer: ```suggestion
As of LLVM 22, Flang has better support for `do concurrent` constructs on the CPU
```

> As of LLVM 22, Flang has better support for do concurrent parallelization on the CPU

Reviewer 2: Can we use a word other than "better" here? It is not clear what we are comparing to. Do you mean, "better than what was in flang before", or "better than some other compilers"?

Reviewer 3: That is meant to compare to previous flang releases.

Would something like be better:
> As of LLVM 22, flang adds more extensive support for parallelizing `do concurrent` loops on the CPU and the GPU. In particular, the `local` specifier, partial support for `reduce`, and automatic mapping of user-defined types are now supported.

and GPU. On the CPU, we validated the feature using FIATS' inference and training

> codes and the results are satisfactory. On the GPU, we have basic support that is

Reviewer 2: Could you specify what is meant by "satisfactory"? It is not clear if it means

- we were able to compile some codes
- we were able to compile most of them
- we were able to compile all of them, but performance is not great

> codes and the results are satisfactory. On the GPU, we have basic support that is

Reviewer 3: Maybe we can phrase it as:
```suggestion
codes where `do concurrent` and OpenMP had very similar acceleration results (for more info see: [1]). On the GPU, we have basic support that is

[1] Automatically Parallelizing Batch Inference on Deep Neural Networks Using Fiats and Fortran 2023 “Do Concurrent” (https://link.springer.com/chapter/10.1007/978-3-032-07612-0_11)
```

Reviewer 2: Thank you. Yes, this is much better. It makes it clear that `DO CONCURRENT` and OpenMP exhibit comparable performance. Some suggested tweaks:

"... using [FIATS](link-to-fiats) where the performance of `DO CONCURRENT` and OpenMP were similar (for more information, see [^1])

[^1] Automatically Parallelizing Batch Inference ..."

Reviewer 2: The suggestions for what has been added in flang-22 are also very good. Thank you.

still in progress; we have offload tests for 1D and 2D saxpy. We also validated

> using codes that do not use extensive user-defined types and/or allocatables.

Reviewer 2: Could we rephrase this to say what users can expect when trying to lower `do concurrent` to GPU's? For instance, it will not work with UDT's or ALLOCATABLE objects?

## Current status

> As of LLVM 22, flang adds more extensive support for parallelizing do concurrent loops

Reviewer 2: For consistency with formatting the formatting of this elsewhere in the document

```suggestion
As of LLVM 22, flang adds more extensive support for parallelizing `do concurrent` loops
```

on the CPU and the GPU. In particular, the local specifier, partial support for reduce,
and automatic mapping of user-defined types are now supported.

> On the CPU, we validated the feature using FIATS' inference and training codes where

Reviewer 2: Adding a link to FIATS may be helpful. I assume that what is linked in the suggestion is what you have used.

```suggestion
On the CPU, we validated the feature using [FIATS](https://github.com/BerkeleyLab/fiats) inference and training codes where
```

> `do concurrent` and OpenMP had very similar acceleration results (for more info see: [1]).

Reviewer 2: nit

```suggestion
`do concurrent` and OpenMP had very similar acceleration results (for more information, see [1]).
```

> On the GPU, we have basic support that is still in progress; we have offload tests for 

Reviewer 2: nit

```suggestion
On the GPU, we have basic support that is still in progress. We have offload tests for 
```

> 1D and 2D saxpy. We also validated using codes that do not use extensive user-defined

Reviewer 2: ```suggestion
1D and 2D saxpy. We also validated using codes that do not make extensive use of user-defined
```

> types and/or allocatables.

Reviewer 2: I assume that both user-defined type and allocatable support is still in progress. Just saying "and" instead of "and/or" may be better.

```suggestion
types and allocatables.
```

[1] Automatically Parallelizing Batch Inference on Deep Neural Networks Using Fiats

> and Fortran 2023 “Do Concurrent”(https://link.springer.com/chapter/10.1007/978-3-032-07612-0_11)

Reviewer 2: ```suggestion
and Fortran 2023 “Do Concurrent” (https://link.springer.com/chapter/10.1007/978-3-032-07612-0_11)
```
